curriculum/challenges/english/01-responsive-web-design/applied-visual-design/create-a-gradual-css-linear-gradient.english.md

```diff
@@ -26,7 +26,7 @@ Use a <code>linear-gradient()</code> for the <code>div</code> element's <code>ba
 ```yml
 tests:
   - text: The <code>div</code> element should have a <code>linear-gradient</code> <code>background</code> with the specified direction and colors.
-    testString: assert(code.match(/background:\s*?linear-gradient\(35deg,\s*?(#CCFFFF|#CFF),\s*?(#FFCCCC|#FCC)\);/gi), 'The <code>div</code> element should have a <code>linear-gradient</code> <code>background</code> with the specified direction and colors.');
+    testString: assert(code.match(/background:\s*?linear-gradient\(\s*?35deg,\s*?(#CCFFFF|#CFF),\s*?(#FFCCCC|#FCC)\s*?\);/gi), 'The <code>div</code> element should have a <code>linear-gradient</code> <code>background</code> with the specified direction and colors.');
```

Comment:
I have been using something like the following to deal with white space where it makes sense and should work for this particular challenge.
```suggestion
    testString: assert($('style').text().replace(/\s/g, '').match(/background:linear-gradient\(35deg,(#CCFFFF|#CFF),(#FFCCCC|#FCC)\);/gi));
```
FYI - You will notice I did not include the assert message above.  That is because for this section and a few other sections, the `text` above the `testString` is used instead.  This allows us to tidy up the challenges as we work on them.

Reply:
That is definitely a lot cleaner, I like it. Tested locally and everything looks good.

Reply:
The only thing I don't like about my suggestion is that someone could technically write the following and pass the tests but the linear-gradient would not display as intended.
```
background: linear-gradient(
  35 deg,
  #ccffff,
  #ffcccc
);
```
Note the space between `35` and `deg`.  The white space replacement would "fix" the incorrect syntax.

Reply:
I guess I didn't think about that. Would doing this be silly?

`assert(code.match(/35deg/) && $('style').text().replace(/\s/g, '').match(/background:linear-gradient\(35deg,(#CCFFFF|#CFF),(#FFCCCC|#FCC)\);/gi));`

Technically the whitespace trim would also fix incorrect hex codes like `# CFF`, so it isn't a real fix for the problem.

Reply:
@lasjorg I came up with another idea which seems to work on Chrome, Edge, and Firefox.  I am not sure if it would work on other browsers though.

The assert would be:
```
assert(getLinearGradientStr().match(/linear-gradient\(35deg,rgb\(204,255,255\),rgb\(255,204,204\)\)/gi));
```
and then we would add the following in the section with id='challengeSeed'
````html
<div id='html-teardown'>

```html
<script>
window.getLinearGradientStr = () => {
  const div = document.querySelector('div');
  const styleProps = getComputedStyle(div);
  for (let prop in styleProps) {
    if (
      prop.indexOf('background') === 0 &&
      styleProps[prop].includes('linear-gradient')
    ) {
      return styleProps[prop].replace(/\s/g, '');
    }
  }
  return '';
}
</script>
```

</div>
````

Reply:
@RandellDawson I like the idea of using computed styles. Can't we just do this using Jquery. The string inside match will have to have the same formatting as the string Jquery returns (spaces after the commas).

`assert($('div').css('background-image').match(/linear-gradient\(35deg, rgb\(204, 255, 255\), rgb\(255, 204, 204\)\)/gi));`

Seems to work as intended. Am I missing something or does using this approach do what we want?

Note: I should say, this will let you use both the shorthand property background and background-image, which I actually think is better. A lot of people do use the actual background-image property when making a linear-gradient (gradients belong to the image data type).

Reply:
Did you test this in multiple browsers like Edge, Safari and Firefox?

Reply:
I have only tested in Firefox and Chrome, but Jquery usually has really good browser support.

`$('div')` in Firefox give me this element which is the first div in the main html outside the #fcc-main-frame iframe

`<div id="MathJax_Message" style="display: none;">`


But in Chrome it gives the div inside the output (inside the iframe). So does using $($0) after inspecting the element so it is selected.

Not sure why they work differently like that.

----
Note: These are the two logs of the difference in background and background-image (In Chrome)

```
$($0).css('background')
"rgba(0, 0, 0, 0) linear-gradient(35deg, rgb(204, 255, 255), rgb(255, 204, 204)) repeat scroll 0% 0% / auto padding-box border-box"
```

```
$($0).css('background-image')
"linear-gradient(35deg, rgb(204, 255, 255), rgb(255, 204, 204))"
```

Reply:
I should also have said that I believe using the short hand (background) is more likely to be inconsistent then using the long form.

> https://api.jquery.com/css/
> 
> Retrieval of shorthand CSS properties (e.g., margin, background, border), although functional with some browsers, is not guaranteed.

Reply:
It is the possibility of inconsistencies which bothers me.  The only true way to really test this is to use snapshots, but that requires tooling which has yet to be added yet.

Reply:
The inconsistency I was referring to only concerns using the shorthand properties.

AFAIK the only valid property to hold a linear-gradient is the background-image property. When you set the gradient using the background shorthand all it does is set it to the background-image property. Just like when you set a color using background it sets the background-color property.

Considering gradients are of data type image I really don't see why there would be browsers that wouldn't report the linear-gradient when querying the background-image property.

Anyway, all the test I did shows no issue. If you can test this locally on Edge and we can get someone to test it on Safari, and they both work, then surely if the test works on all major browsers it is "good enough".

----
Some specs links just as reference.

[CSS Image Values and Replaced Content Module Level 3: gradients](https://www.w3.org/TR/css3-images/#gradients)

[CSS Cascading and Inheritance Level 4: shorthand](https://drafts.csswg.org/css-cascade-4/#shorthand)

Reply:
@raisedadead What do you think about this approach?  We won't really know until we push the changes up to freecodecamp.rocks and maybe put a post out on the forum and request people with various browsers test out the challenge.  I like the concept behind this, but we don't want to flood the forum with new questions or have people start creating issues when the have written valid code but their particular browser is not supported by jQuery when using css('background-image').  I believe @lasjorg is right about the `background-image` being more likely to return the linear gradient, but will it always look the same as far as the format?

Reply:
> [...] We won't really know until we push the changes up to freecodecamp.rocks and maybe put a post out on the forum and request people with various browsers test out the challenge [...]

Why can't we test this locally? I mean why do we require this fix to be deployed first on the beta and then confirm?

Reply:
Because I don't have access to certain browsers for testing.

Reply:
I used lambdatest screenshot feature with a Codepen that prints the string to the screen. It is the best I can do, without a paid account the live testing is limited.

Here are the images:

windows-8-ie-10
![windows-8-ie-10-keqnb](https://user-images.githubusercontent.com/28780271/53685851-87e9b900-3d20-11e9-8370-50999be5d13b.png)
windows-10-chrome-71
![windows-10-chrome-71-zvda6](https://user-images.githubusercontent.com/28780271/53685852-88824f80-3d20-11e9-827e-7184ee65192b.png)
windows-10-ie-11
![windows-10-ie-11-5ffps](https://user-images.githubusercontent.com/28780271/53685853-88824f80-3d20-11e9-810d-eda04fa0eff5.png)
windows-10-mozilla-64
![windows-10-mozilla-64-rpqqd](https://user-images.githubusercontent.com/28780271/53685854-88824f80-3d20-11e9-83ad-89cb9002a3cb.png)
windows-10-opera-56
![windows-10-opera-56-syp4t](https://user-images.githubusercontent.com/28780271/53685855-88824f80-3d20-11e9-95af-28e6b45d1fcb.png)
windows-xp-ie-8 (fail Codepen issue i think)
![windows-xp-ie-8-knxsn](https://user-images.githubusercontent.com/28780271/53685856-891ae600-3d20-11e9-8965-488258078281.png)
android-8 1-galaxy-note-9
![android-8 1-galaxy-note-9-djjkt](https://user-images.githubusercontent.com/28780271/53685857-891ae600-3d20-11e9-9909-d9e4e7cd06dd.png)
ios-10 3-ipad-pro
![ios-10 3-ipad-pro-lr3mb](https://user-images.githubusercontent.com/28780271/53685858-891ae600-3d20-11e9-8d50-1552c76ae69b.png)
ios-10 3-iphone-7
![ios-10 3-iphone-7-u4pzi](https://user-images.githubusercontent.com/28780271/53685859-891ae600-3d20-11e9-8152-f521674ddf88.png)
macos-mojave-safari-12
![macos-mojave-safari-12-sbirh](https://user-images.githubusercontent.com/28780271/53685860-891ae600-3d20-11e9-9a9c-f271018430d6.png)
windows-7-ie-11
![windows-7-ie-11-5wjgi](https://user-images.githubusercontent.com/28780271/53685861-891ae600-3d20-11e9-8dd1-7dc362c7a730.png)
windows-8 1-ie-11
![windows-8 1-ie-11-cuevk](https://user-images.githubusercontent.com/28780271/53685862-89b37c80-3d20-11e9-84c9-6ce44a98b666.png)

-----

Note: I have compared the standard API against Jquery and the return string is the same. I think the Jquery formatting is just the web API. I don't think Jquery does any formating of the return string. After a quick peek at the Jquery source it looks to me like they do this (Sizzle is obviously not just using querySelector).


```
const element = document.querySelector('.lg');
const computed = window.getComputedStyle(element);
const propVal = computed.getPropertyValue('background-image');

```
Source:

https://github.com/jquery/jquery/blob/master/src/css/var/getStyles.js
https://github.com/jquery/jquery/blob/master/src/css/curCSS.js


So I think if there are inconsistencies they would show up with "vanilla" JS as well.

Reply:
@lasjorg So it appears use `background-image` in the selector should work.  Great job in fleshing this out!  I could not tell from the screenshots, but were there any spacing differences where we should remove whitespace from the value returned just in case?

Reply:
I too really only have the screenshots to go by. I was able to test a few browsers in live mode and for the ones I did test they all looked the same. As do the screenshots to me. A third pair of eyes would be good, but I'm pretty sure they are all the same formatting. I don't think we need to trim any whitespace.

Reply:
Running test again with length test. One sec.

Reply:
On a related note. Maybe FCC should ask one of the on-line browser testing sites for a sponsored deal so we can run better cross browser testing?

These guys have some deals with open source projects.
https://crossbrowsertesting.com/open-source

Not sure if other sites would be interested but it seems like it would be pretty good exposure for them.

Reply:
@lasjorg Go ahead and update the tests using the `background-image` to capture the linear-gradient value for testing purposes.

Reply:
Do we need to test for a semicolon at the end?

Reply:
@lasjorg If the linear-gradient is the last property in the selector definition, then technically it is not needed.  However, if someone puts the linear-gradient anywhere else in the definition and there are other properties below it, then it will not render correctly and the other tests will fail.  I think not requiring it here is fine, because if someone leaves it off when the linear-gradient is at the end, the test should pass (and it does and renders correctly).  If someone leaves it off and puts the linear-gradient above other properties (the test should fail because the linear-gradient will not render).